covid19-dashboard_NEW/calculate_data.py

```diff
@@ -0,0 +1,173 @@
+    total: int = sum(value for key, value in values[:chunk])
+
+    output: dict = {}
+
+    # We move in sizes of chunk_size, storing the sum from n to n+chunk_size
+    # on output[date]. Every movement, we subtract the first element
+    # and add the last element. Similar to how a queue works,
+    # remove the first add to end.
+    for i in range(chunk, len(values)):
+        date, value = values[i]
+        # Add the current value and subtract the oldest value from the list.
+        total += value - values[i - chunk][1]
+        # Get the current average.
+        average: float = total / chunk
+        # Store the average for this date under output[date].
+        output[date] = average
+
+    return output
+
+
+def _clean_dataset(date_to_val: dict,
+                   step_size: int = 2,
+                   start_date: str = '2020-03-01'):
+    """
+    Given a dataset, return the elements whose index is divisible by step_size.
+    Remove any negative values and dates starting prior to start_date.
+    :param start_date: any date before this date will be omitted.
+    :param date_to_val: an object of type date->int.
+    :param step_size: the size of our steps.
+    :return: the same dataset but 1/nth of the size.
+    """
+    # The latest date is the largest date from the list of dates.
+    latest_date: str = max(list(date_to_val.keys()))
+
+    # Get a list of (Key, Value).
+    # Key = date, Value = cases/deaths.
+    date_to_val_list = list(date_to_val.items())
+    # Sort the values by date, elem[0].
+    date_to_val_list = sorted(date_to_val_list, key=lambda e: e[0])
+
+    # Only include the element if its index is divisible by chunk.
+    date_to_val_list = date_to_val_list[::step_size]
+
+    # Remove dates before March 1st.
+    date_to_val_list = [
+        (date, value)
+        for date, value in date_to_val_list
+        if date >= start_date
+    ]
+
+    # Remove negative values from the list.
+    date_to_val_list = [
+        (date, 0) if value < 0 else (date, value)
+        for date, value in date_to_val_list
+    ]
+
+    # Because the latest date might have been skipped by the step_size.
+    # Check to see if it's in our dataset, otherwise append it.
+    # We need the latest date to be present.
+    if date_to_val_list[-1][0] != latest_date:
+        date_to_val_list.append((latest_date, date_to_val[latest_date]))
+
+    # Convert the list of (date, value) back to a dictionary.
+    output = dict(date_to_val_list)
+
+    return output
+
+
+def calculate_data(data_type: str, data: dict, population: int) -> dict:
+    """
+    Given a dataset for a region and its population,
+    perform different types of calculations on every date.
+    :param data_type: type of data, "cases" or "deaths".
+    :param data: a dictionary containing date->int.
+    :param population: geoId -> population.
+    :return: a dictionary with the 7-week moving average + other calculations.
```

Comment:
please be explicit about the structure of the return dict.

Reply:
Fixed. Added return type, README.md info and more comments.